### ISSUE TITLE
Launching container on tagged worker takes forever / 30min+

### ISSUE THREAD

**Issue by drahnr** [concourse/concourse#5334]
## Bug Report

Tagged worker launches _very_ delayed

## Steps to Reproduce

```ini
───────┬─────────────────────────────────────
       │ File: /etc/concourse/garden.ini
───────┼─────────────────────────────────────
   1   │ [server]
   2   │ network-pool = 172.16.0.0/16
   3   │ dns-server = 127.0.0.1
   4   │ dns-server = 1.1.1.1
   5   │ runtime-plugin = /usr/bin/nvidia-container-runtime
───────┴─────────────────────────────────────

```

excerpt of a task failing
```yaml
   - name: test-juice
     build_logs_to_retain: 4
     public: true
     serial: true
     plan:
     - in_parallel:
       - get: juice
         trigger: true
       - get: container-fedora-cuda
         trigger: true
 
       - get: container-fedora-default
         trigger: true
 
       - get: container-fedora-native
         trigger: true
 
     - task: build-fedora-cuda
       image: container-fedora-cuda
       tags: [framework:cuda]
       #privileged: true # not needed since runc has a nvidia container image plugin
       # but does not make a difference anyways
       config:
         platform: linux
         inputs:
         - name: juice
         run:
           path: sh
           args:
           - -exc
           - |
             prepare
             cd cudnn-sys
             cargo build
             cargo build --tests
             cd -
             cd cudnn
             cargo build
             cargo build --tests
           dir: juice/rust-cudnn
```

with a concourse worker and tags `--tag vendor:nvidia --tag framework:opencl --tag framework:cuda`

as well as using the sysem native (fedora 31) `runc`

## Expected Results

Just work like a normal worker.
I should see some activity in `gdn`, also extracting the `gdn` into it's own native service unit and launching with debug does not show any additional information, same for running `runc` with `--debug` via `--runtime-plugin=/usr/bin/runc` and `--runtime-plugin-extra-arg="--debug"`

## Actual Results

Tasks never get executed or very delayed, i.e. https://ci.spearow.io/teams/spearow/pipelines/juice/jobs/test-juice/builds/63

## Additional Context
Note that the garden and concourse worker do not show any logs while the web already shows resource checks being all checked (green check marks), but the task itself is showing as yellow. I still do not see any messages from either worker or garden. 

## Version Info

* Concourse version:
5.4.1
5.8.0
* Deployment type (BOSH/Docker/binary):
binary/docker
* Infrastructure/IaaS:
Workers: Fedora 31
Web: Docker
* Browser (if applicable):
Firefox, not applicable
* Did this used to work?
Yes, but that was 3.x

**Comment by drahnr** [comment]
Sample is here - between one task finishes and another starts running, it takes a whoping 35minutes https://ci.spearow.io/teams/spearow/pipelines/juice/jobs/pr-test-juice/builds/119

**Comment by drahnr** [comment]
Still present in v6.0.0

![Screenshot_2020-03-27 build-pkg-rpm #35 - Concourse](https://user-images.githubusercontent.com/667047/77739824-80274180-7012-11ea-9ea5-213b7f21c315.png)

Pipeline:
https://github.com/drahnr/wondershaper-rpm/blob/master/.concourse.yml

Deployed Pipeline:
https://ci.spearow.io/teams/main/pipelines/wondershaper


Addendum information:

I have one worker which is tagged with `concourse worker <snip> --tag framework:cuda --tag framework:opencl --tag framework vendor:nvidia<snip>` and one without tags.
`garden`/`gdn` uses `containerd` as backend and for process management.

**Comment by cirocosta** [comment]
Hey @drahnr , would you mind sharing how you're deploying that worker, more specifically, with regards to how you're customizing the runtime (i.e., passing those flags to runc, etc)?

Also, one thing to note is that Concourse needs to stream that input to your task (as the `get` happens in a non-tagged worker, in your example). Is `juice` something that's very large and thus slow to stream?

thanks!

**Comment by drahnr** [comment]
@cirocosta thanks for looking into it!

`nvidia-container-runtime` is essentially a shim modifying the config json fed into runc, but it launches runc itself, so there should be no difference to only running `runc` (see https://github.com/NVIDIA/nvidia-container-runtime )


```
───────┬──────────────────────────────────────────────────────────────────
       │ File: /etc/containers/oci/hooks.d/nvidia-tools.json
───────┼──────────────────────────────────────────────────────────────────
   1   │ {
   2   │   "version": "1.0.0",
   3   │   "hook": {
   4   │     "path": "/usr/bin/nvidia-container-runtime-hook",
   5   │     "args": ["prestart"],
   6   │     "env": [
   7   │       "NVIDIA_REQUIRE_CUDA=cuda>=8.0",
   8   │       "NVIDIA_VISIBLE_DEVICES=all",
   9   │       "NVIDIA_DRIVER_CAPABILITIES=video,graphics,compute,utility"
  10   │     ]
  11   │   },
  12   │   "when": {
  13   │     "always": true
  14   │   },
  15   │   "stages": ["prestart"]
  16   │ }
───────┴──────────────────────────────────────────────────────────────────
```

**Comment by vito** [comment]
@drahnr It might be worth configuring [tracing](https://concourse-ci.org/tracing.html) (new in v6.0) - it should show exactly where this time is being spent, assuming we've put spans in the right places.

You might also want to consider configuring `inputs: detect` or `inputs: [foo, bar]` so that Concourse doesn't stream all the assets to the `put` step. This will limit the `put` step to only the assets that it needs.

**Comment by drahnr** [comment]
So I did not get around to trying `jaeger`, but I added a few more monitoring checks as  well as adding sample jobs which use the cross tag/non-tag references of resources/tasks - and the issue reproduces nicely for untagged resources and tagged tasks (as well as vice versa).
But as it seems another issue is causing another delay.
In my main pipeline I moved a few things around for more determinism:

* replace `docker-image` with `registry-image`
* make sure all tasks only use matching tagged/untagged resources

I am going to check if the root cause here is that different tasks within one job require different tags.
Could that be an issue?
According to `https://concourse-ci.org/jobs.html#schema.step.tags` that would mean all jobs are placed at the properly tagged / untagged worker where the resources should already exist.
Does the "one-job" frame cause issues here?

**Comment by Alphasite** [comment]
Since your image is large, you may want to try using btrfs as the overlay file system, overlay is very slow. We ran into a very similar issue where the underlying cause appeared to be starting multiple containers on the machine would saturate the IOPs for the underlying datastore and cause the container creation to grind to a crawl (or with enough parallel jobs hit the 2h creation timeout).

**Comment by drahnr** [comment]
@Alphasite I set the drivers for both concourse as well as containerd via /etc/containers/storage.conf to be btrfs

**Comment by drahnr** [comment]
This issue was resolved along with the streaming images issue #5517 and can not be reproduced wtih `v6.5.1`.
